### PR TITLE
ci: add debug_asan_clang workflow

### PR DESCRIPTION
**Pull request by nshy**
Similarly to release_asan_clang but to test debug build. It is also run only under `asan-ci` and `full-ci` labels.

Fiber stack size is 2 times bigger than in the release workflow for luajit tests to pass. Note that this factor is a wild guess.

Part of #7327

### CODE REVIEW

File: .test.mk [hunk 1]
```diff
@@ -123,6 +123,36 @@ test-release-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.
                                                   unmap_shadow_on_exit=1
 test-release-asan: build run-luajit-test run-test
 
+# Debug ASAN build
+
+.PHONY: test-debug-asan
+test-debug-asan: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=Debug \
+                                  -DENABLE_WERROR=ON \
```
Comment (ylobankov) [inline .test.mk:130]: Please align it with the `-DCMAKE_BUILD_TYPE=Debug. It should be one column.

File: .test.mk [hunk 2]
```diff
@@ -123,6 +123,36 @@ test-release-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.
...
+                                  -DENABLE_ASAN=ON \
+                                  -DENABLE_UB_SANITIZER=ON \
+                                  -DFIBER_STACK_SIZE=1280Kb \
```
Comment (ylobankov) [inline .test.mk:133]: You should leave a comment about increasing timeouts.

File: .test.mk [hunk 3]
```diff
@@ -123,6 +123,36 @@ test-release-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.
...
+                                             print_suppressions=0
+test-debug-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
+                                   ASAN_OPTIONS=detect_invalid_pointer_pairs=1:$\
+                                                detect_leaks=1:$\
+                                                dump_instruction_bytes=1:$\
+                                                heap_profile=0:$\
+                                                print_suppressions=0:$\
+                                                symbolize=1:$\
+                                                unmap_shadow_on_exit=1
+test-debug-asan: TEST_RUN_PARAMS += --test-timeout 620 \
```
Comment (ylobankov) [inline .test.mk:151]: You should leave a comment about increasing timeouts.

File: src/lua/fiber.lua [hunk 4]
```diff
@@ -24,19 +25,24 @@ Sets the default value for the max fiber slice. The old value is infinity
 https://tarantool.io/compat/fiber_slice_default
 ]]
 
+local max_slice_default
+if tarantool.build.asan then
+    max_slice_default = TIMEOUT_INFINITY
```
Comment (locker) [inline src/lua/fiber.lua:30]: Please add a comment explaining why this is required.

File: .test.mk [hunk 5]
```diff
@@ -123,6 +123,36 @@ test-release-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.
...
+                                             detect_leaks=1:dump_instruction_bytes=1:$\
+                                             print_suppressions=0
+test-debug-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
+                                   ASAN_OPTIONS=detect_invalid_pointer_pairs=1:$\
+                                                detect_leaks=1:$\
+                                                dump_instruction_bytes=1:$\
+                                                heap_profile=0:$\
+                                                print_suppressions=0:$\
+                                                symbolize=1:$\
+                                                unmap_shadow_on_exit=1
```
Comment (locker) [inline .test.mk:150]: A lot of code duplication with `test-release-asan`. Can you share the common parts?

Comment (nshy) [reply .test.mk:150]: Done.

<details>
<summary>diff</summary>

```diff
diff --git a/.test.mk b/.test.mk
index 2d927353d..ed12442d2 100644
--- a/.test.mk
+++ b/.test.mk
@@ -81,6 +81,35 @@ test-perf: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=RelWithDebInfo \
 
 test-perf: build run-perf-test
 
+# *_ASAN variables are common part of respective variables for all ASAN builds.
+CMAKE_PARAMS_ASAN = -DENABLE_WERROR=ON \
+                    -DENABLE_ASAN=ON \
+                    -DENABLE_UB_SANITIZER=ON \
+                    -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=OFF \
+                    -DENABLE_FUZZER=ON \
+                    -DTEST_BUILD=ON
+# Some checks are temporary suppressed in the scope of the issue
+# https://github.com/tarantool/tarantool/issues/4360:
+#   - ASAN: to suppress failures of memory error checks caught while tests run, the asan/asan.supp
+#     file is used. It is set as a value for the `-fsanitize-blacklist` option at the build time in
+#     the cmake/profile.cmake file.
+#   - LSAN: to suppress failures of memory leak checks caught while tests run, the asan/lsan.supp
+#     file is used.
+TEST_RUN_ENV_ASAN = ASAN=ON \
+                    LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
+                    ASAN_OPTIONS=heap_profile=0:unmap_shadow_on_exit=1:$\
+                                 detect_invalid_pointer_pairs=1:symbolize=1:$\
+                                 detect_leaks=1:dump_instruction_bytes=1:$\
+                                 print_suppressions=0
+LUAJIT_TEST_ENV_ASAN = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
+                       ASAN_OPTIONS=detect_invalid_pointer_pairs=1:$\
+                                    detect_leaks=1:$\
+                                    dump_instruction_bytes=1:$\
+                                    heap_profile=0:$\
+                                    print_suppressions=0:$\
+                                    symbolize=1:$\
+                                    unmap_shadow_on_exit=1
+
 # Release ASAN build
 
 .PHONY: test-release-asan
@@ -92,62 +121,21 @@ test-perf: build run-perf-test
 # while running LuaJIT tests with ASan support enabled.
 # Experiments once again confirm the notorious quote that "640 Kb
 # ought to be enough for anybody".
-test-release-asan: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=RelWithDebInfo \
-                                  -DENABLE_WERROR=ON \
-                                  -DENABLE_ASAN=ON \
-                                  -DENABLE_UB_SANITIZER=ON \
-                                  -DFIBER_STACK_SIZE=640Kb \
-                                  -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=OFF \
-                                  -DENABLE_FUZZER=ON \
-                                  -DTEST_BUILD=ON
-# Some checks are temporary suppressed in the scope of the issue
-# https://github.com/tarantool/tarantool/issues/4360:
-#   - ASAN: to suppress failures of memory error checks caught while tests run, the asan/asan.supp
-#     file is used. It is set as a value for the `-fsanitize-blacklist` option at the build time in
-#     the cmake/profile.cmake file.
-#   - LSAN: to suppress failures of memory leak checks caught while tests run, the asan/lsan.supp
-#     file is used.
-test-release-asan: TEST_RUN_ENV = ASAN=ON \
-                                  LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
-                                  ASAN_OPTIONS=heap_profile=0:unmap_shadow_on_exit=1:$\
-                                               detect_invalid_pointer_pairs=1:symbolize=1:$\
-                                               detect_leaks=1:dump_instruction_bytes=1:$\
-                                               print_suppressions=0
-test-release-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
-                                     ASAN_OPTIONS=detect_invalid_pointer_pairs=1:$\
-                                                  detect_leaks=1:$\
-                                                  dump_instruction_bytes=1:$\
-                                                  heap_profile=0:$\
-                                                  print_suppressions=0:$\
-                                                  symbolize=1:$\
-                                                  unmap_shadow_on_exit=1
+test-release-asan: CMAKE_PARAMS = ${CMAKE_PARAMS_ASAN} \
+                                  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+                                  -DFIBER_STACK_SIZE=640Kb
+test-release-asan: TEST_RUN_ENV = ${TEST_RUN_ENV_ASAN}
+test-release-asan: LUAJIT_TEST_ENV = ${LUAJIT_TEST_ENV_ASAN}
 test-release-asan: build run-luajit-test run-test
 
 # Debug ASAN build
 
 .PHONY: test-debug-asan
-test-debug-asan: CMAKE_PARAMS = -DCMAKE_BUILD_TYPE=Debug \
-                                  -DENABLE_WERROR=ON \
-                                  -DENABLE_ASAN=ON \
-                                  -DENABLE_UB_SANITIZER=ON \
-                                  -DFIBER_STACK_SIZE=1280Kb \
-                                  -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=OFF \
-                                  -DENABLE_FUZZER=ON \
-                                  -DTEST_BUILD=ON
-test-debug-asan: TEST_RUN_ENV = ASAN=ON \
-                                LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
-                                ASAN_OPTIONS=heap_profile=0:unmap_shadow_on_exit=1:$\
-                                             detect_invalid_pointer_pairs=1:symbolize=1:$\
-                                             detect_leaks=1:dump_instruction_bytes=1:$\
-                                             print_suppressions=0
-test-debug-asan: LUAJIT_TEST_ENV = LSAN_OPTIONS=suppressions=${PWD}/asan/lsan.supp \
-                                   ASAN_OPTIONS=detect_invalid_pointer_pairs=1:$\
-                                                detect_leaks=1:$\
-                                                dump_instruction_bytes=1:$\
-                                                heap_profile=0:$\
-                                                print_suppressions=0:$\
-                                                symbolize=1:$\
-                                                unmap_shadow_on_exit=1
+test-debug-asan: CMAKE_PARAMS = ${CMAKE_PARAMS_ASAN} \
+                                -DCMAKE_BUILD_TYPE=Debug \
+                                -DFIBER_STACK_SIZE=1280Kb
+test-debug-asan: TEST_RUN_ENV = ${TEST_RUN_ENV_ASAN}
+test-debug-asan: LUAJIT_TEST_ENV = ${LUAJIT_TEST_ENV_ASAN}
 test-debug-asan: TEST_RUN_PARAMS += --test-timeout 620 \
                                     --no-output-timeout 630 \
                                     --server-start-timeout 610

```

</details>

File: .test.mk (hunk 1 above)
Comment (nshy) [reply .test.mk:130]: Done. See the diff in the comment below about deduplication.

File: .test.mk (hunk 2 above)
Comment (nshy) [reply .test.mk:133]: Done.

```diff
diff --git a/.test.mk b/.test.mk
index ed12442d2..dd81c17ad 100644
--- a/.test.mk
+++ b/.test.mk
@@ -136,6 +136,8 @@ test-debug-asan: CMAKE_PARAMS = ${CMAKE_PARAMS_ASAN} \
                                 -DFIBER_STACK_SIZE=1280Kb
 test-debug-asan: TEST_RUN_ENV = ${TEST_RUN_ENV_ASAN}
 test-debug-asan: LUAJIT_TEST_ENV = ${LUAJIT_TEST_ENV_ASAN}
+# Increase timeouts as some tests in ASAN debug build take quite a lot
+# of time to finish.
 test-debug-asan: TEST_RUN_PARAMS += --test-timeout 620 \
                                     --no-output-timeout 630 \
                                     --server-start-timeout 610
```

File: .test.mk (hunk 3 above)
Comment (nshy) [reply .test.mk:151]: Done.

```diff
diff --git a/.test.mk b/.test.mk
index ed12442d2..dd81c17ad 100644
--- a/.test.mk
+++ b/.test.mk
@@ -136,6 +136,8 @@ test-debug-asan: CMAKE_PARAMS = ${CMAKE_PARAMS_ASAN} \
                                 -DFIBER_STACK_SIZE=1280Kb
 test-debug-asan: TEST_RUN_ENV = ${TEST_RUN_ENV_ASAN}
 test-debug-asan: LUAJIT_TEST_ENV = ${LUAJIT_TEST_ENV_ASAN}
+# Increase timeouts as some tests in ASAN debug build take quite a lot
+# of time to finish.
 test-debug-asan: TEST_RUN_PARAMS += --test-timeout 620 \
                                     --no-output-timeout 630 \
                                     --server-start-timeout 610
```

File: src/lua/fiber.lua (hunk 4 above)
Comment (nshy) [reply src/lua/fiber.lua:30]: Done. 

```diff
--- a/src/lua/fiber.lua
+++ b/src/lua/fiber.lua
@@ -25,6 +25,8 @@ Sets the default value for the max fiber slice. The old value is in>
 https://tarantool.io/compat/fiber_slice_default
 ]]

+-- ASAN build is slow. Turn slice check off to suppress noisy failures
+-- on exeeding slice limit in this case.
 local max_slice_default
 if tarantool.build.asan then
     max_slice_default = TIMEOUT_INFINITY
```

Comment (locker) [reply src/lua/fiber.lua:32]: I'd added the same comment to `fiber.c` as well.

**Review by locker — APPROVED** [review]

File: src/lua/fiber.lua (hunk 4 above)
Comment (nshy) [reply src/lua/fiber.lua:32]: Done.

```diff
diff --git a/src/lib/core/fiber.c b/src/lib/core/fiber.c
index 2fda2eec9..611abaa11 100644
--- a/src/lib/core/fiber.c
+++ b/src/lib/core/fiber.c
@@ -50,6 +50,11 @@
 extern void cord_on_yield(void);

 static struct fiber_slice zero_slice = {.warn = 0.0, .err = 0.0};
+
+/**
+ * ASAN build is slow. Turn slice check off to suppress noisy failures
+ * on exceeding slice limit in this case.
+ */
 #if ENABLE_ASAN
 static struct fiber_slice default_slice = {
        .warn = TIMEOUT_INFINITY,
@@ -58,6 +63,7 @@ static struct fiber_slice default_slice = {
 #else
 static struct fiber_slice default_slice = {.warn = 0.5, .err = 1.0};
 #endif
+
 /** Number of cord-threads still running right now. */
 static int cord_count = 0;
```

**Review by ylobankov — APPROVED** [review]

**Comment by locker** [comment]
@ylobankov Do you think we should backport this PR to 2.11 and 2.10?

If yes, @nshy please create separate PRs for 2.11 and 2.10.

**Comment by ylobankov** [comment]
> @ylobankov Do you think we should backport this PR to 2.11 and 2.10?
> 
> If yes, @nshy please create separate PRs for 2.11 and 2.10.

If it can be easily backported, then let's do it.

**Pull request merged**